File: solutions/053_permute.go
package main

func permute(nums []int) [][]int {
	rtv := [][]int{}
	var backtrace func([]int, []int)
	// options代表可供填入数字 permute是已填入的数字组合
	backtrace = func(options, permute []int) {
		if len(options) == 0 {
			rtv = append(rtv, permute)
			return
		}
		for i := 0; i < len(options); i++ {
			newOptions := []int{}
			for j := 0; j < len(options); j++ {
				if j != i {
					newOptions = append(newOptions, options[j])
				}
			}
			backtrace(newOptions, append(permute, options[i]))
		}
	}
	backtrace(nums, []int{})
	return rtv
}

File: solutions/__053_permute.cpp
/*
 *      Author: peanutzhen
 *      Created time: 2021/9/1 21:19:00
 */

#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> ans;
int n;

void backtrace(vector<int> &nums, int depth) {
    if (depth == n) {
        ans.push_back(nums);
        return;
    }
    for (int i = depth; i < n; ++i) {
        swap(nums[depth], nums[i]);
        backtrace(nums, depth + 1);
        swap(nums[depth], nums[i]);
    }
}

vector<vector<int>> permute(vector<int> &nums) {
    ans.clear();
    n = (int) nums.size();
    backtrace(nums, 0);
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);


    return 0;
}

